man/man3/pmsetmode.3
'\"macro stdmacro
.\"
.\" Copyright (c) 2016 Red Hat.
.\" Copyright (c) 2000-2004 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMSETMODE 3 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmSetMode\f1 \- set collection time and mode parameters for the current PMAPI context
.SH "C SYNOPSIS"
.ft 3
.ad l
.hy 0
#include <pcp/pmapi.h>
.sp
int pmSetMode(int \fImode\fP,
'in +\w'int pmSetMode('u
const\ struct\ timespec\ *\fIwhen\fP,
const\ struct\ timespec\ *\fIdelta\fP);
.in
.sp
cc ... \-lpcp
.hy
.ad
.ft 1
.SH DESCRIPTION
.de CR
.ie t \f(CR\\$1\f1\\$2
.el \fI\\$1\f1\\$2
..
.B pmSetMode
are used to define the collection time and/or mode for accessing
performance metrics and metadata in the current
Performance Metrics Application Programming Interface (PMAPI)
context.
This mode affects the semantics of subsequent calls to the following
PMAPI routines:
.BR pmFetch (3),
.BR pmFetchArchive (3),
.BR pmLookupDesc (3),
.BR pmGetInDom (3),
.BR pmLookupInDom (3),
.BR pmLookupLabels (3)
and
.BR pmNameInDom (3).
.PP
Intended mainly for retrospective analysis of performance metrics
from a PCP archive, the
options described below allow an application to implement seeking to an arbitrary
time within the archive, playback, fast forward, reverse,
etc. by alternating calls
to
.B pmSetMode
and
.BR pmFetch (3).
.PP
If
.I mode
is
.B PM_MODE_LIVE
then all information is returned from the active pool of performance metrics
as of the time that the PMAPI call is made, and the other two parameters to
.B pmSetMode
(\c
.I when
and
.IR delta )
are ignored.
.B PM_MODE_LIVE
is the default mode when a new PMAPI context is created with type
.B PM_CONTEXT_HOST
(i.e. fetching metrics from
.B pmcd (1))
or
.BR PM_CONTEXT_LOCAL ,
and this
.I mode
is rarely used in calls to
.B pmSetMode
.PP
The other values of
.I mode
are used with PCP archives where the associated PMAPI context must be of type
.B PM_CONTEXT_ARCHIVE
(when it was created
with
.BR pmNewContext (3))
and the
.I when
parameter defines the current time within the archive.
All requests for metric values and metadata (metric
descriptions and instance identifiers from the instance domains) will be
processed to reflect the state in the archive as of the current time.
.PP
When selecting a value for
.IR when ,
the time at the start of the archive can be established from
the
.CR ll_start
field in the structure returned from a call to
.BR pmGetArchiveLabel (3)
or
the
.CR start
field in the structure returned from a call to
.BR pmGetArchiveLabel (3),
and the time at the end of the archive can be established by calling
.BR pmGetArchiveEnd (3)
.PP
As a special case, if
.I when
is
.B NULL
then the
.I mode
and
.I delta
arguments are used as described below, but the current time in the archive
is not altered.
.SH "TIME-DRIVEN ARCHIVE MODE"
If
.I mode
is
.B PM_MODE_INTERP
then as metric values are retrieved from the archive with
.BR pmFetch (3)
the current time is returned as the timestamp in the
.CR pmResult
structure and
the current time moves on; if
.I delta
is positive, the current time moves forwards, else the current
time moves backwards.
The adjustment to the current time is applied even if the
.BR pmFetch (3)
fails to return values for any metrics or returns an error,
e.g. \c
.B PM_ERR_EOL
because the current time is outside the range defined by
the records in the archive.
.PP
When metric values are being requested via
.BR pmFetch (3)
the current time may
not exactly match the times at which values have been recorded in the
archive, so the returned metric values
are computed from the observed metric values
in the archive (usually at times close to the current time).
.PP
For metrics with semantics of
.BR PM_SEM_COUNTER ,
the computed value is based
on linear interpolation between the last observation before
the current time and the first
observation after the current time.
.PP
For metrics with semantics of
.B PM_SEM_INSTANT
or
.BR PM_SEM_DISCRETE ,
the computed value is based on values in the neighbourhood of the current time.
.PP
The algorithms used in these computations depend on the semantics
of the metrics and the time series of observed values in the archive; a fuller
explanation may be found in the white paper
.I "Explaining Value Interpolation with PCP Archives"
found at
.BR https://pcp.io/papers/archive-interpolation.pdf .
.SH "RECORD-DRIVEN ARCHIVE MODES"
If
.I mode
is
.B PM_MODE_FORW
or
.B PM_MODE_BACK
then when metric values are being requested via
.BR pmFetch (3)
the archive will be scanned in a forwards (\c
.BR PM_MODE_FORW )
or backwards (\c
.BR PM_MODE_BACK )
direction in time, until an archive record is
encountered with values for
.B "at least one"
of the requested metrics.
This archive record is used to provide as many of
the requested metrics as possible and these are
returned with the timestamp if the record in the archive, which becomes the
new current time.
.PP
Note that any metrics requested via
.BR pmFetch (3)
that do
.B not
have a value in the archive record at the
new current time will return no values, and so this mode is most useful for
archives where all of the metrics of interest have been logged regularly
at the same time in the archive.
Otherwise, each
.BR pmFetch (3)
will contain only the subset of the requested metrics and any associated
instances found in the
qualifying archive record.
.PP
The
.I delta
parameter is ignored, because the current time is driven by the
timestamp in the matching archive record.
So there is no concept
of stepping through the archive in regular time with this mode,
although if the requested metrics appear at regular intervals
in the archive the current time may advance by regular intervals,
but this is serendipitous.
.PP
If no qualifying metrics can be found in the requested direction of searching
before the end or start of the archive is encountered, then
.BR pmFetch (3)
returns the special error indicator,
.BR PM_ERR_EOL .
.SH RECOMMENDATIONS
When processing PCP archives,
.B PM_MODE_INTERP
is preferred because:
.IP \(bu 2n
This maximizes the information that will be returned
in each
.BR pmFetch (3)
.IP \(bu 2n
This returns values at regular intervals of the current time, independent
of the logging frequency for requested metrics in the archive.
.IP \(bu 2n
This works with
.B any
PCP archive, as opposed to the record-driven modes which may work
acceptably for archives with regular logging of all requested metrics,
but may fail to report complete or useful results for other archives.
.IP \(bu 2n
This mode provides the closest semantic match to
.B PM_MODE_LIVE
and leads to the least user surprise when moving between real-time
monitoring and retrospective analysis.
.SH EXAMPLES
To replay interpolated metrics from an archive at 10 second intervals,
the following code fragment could be used:
.PP
.ft CR
.nf
.in +0.5i
struct timeval mytime;
int mydelta = 10 * 1000;      /* msec */
pmLogLabel label;
pmResult result;

pmNewContext(PM_CONTEXT_ARCHIVE, "myarchive");
pmGetArchiveLabel(&label);
mytime = label.ll_start;
pmSetMode(PM_MODE_INTERP, &mytime, mydelta)

while (pmFetch(numpmid, pmidlist, &result) != PM_ERR_EOL) {
    /*
     * process interpolated metric values as of
     * result->timestamp
     */
    \&. . .
    pmFreeResult(result);
}
.in
.fi
.ft 1
.PP
The following code fragment may be used to dump values
for selected metrics in an
archive in reverse temporal sequence.
.PP
.ft CR
.nf
.in +0.5i
struct timespec mytime;
pmResult result;

pmNewContext(PM_CONTEXT_ARCHIVE, "myarchive");
pmGetArchiveEnd(&mytime);
pmSetMode(PM_MODE_BACK, &mytime, NULL);

while (pmFetch(npmid, pmidlist, &result) != PM_ERR_EOL) {
    /*
     * process logged metric values as of result->timestamp
     */
    \&. . .
    pmFreeResult(result);
}
.in
.ft 1
.SH COMPATIBILITY
Prior to PCP 7.0 the
.I when
argument was a \f(CRstruct timeval\fP
and the
.I delta
argument was an \f(CRint\fP (in units of milliseconds).
To support PMAPI transition, the old interface and semantics can be
used if applications are recompiled with
.BR \-DPMAPI_VERSION=2 .
.PP
For a time in PCP 6.x there was a
routine with the same semantics as the current
.B pmSetMode
called
.B pmSetModeHighRes
although this is now deprecated and compile-time support for
.B pmSetModeHighRes
will be removed in a future release.
.SH DIAGNOSTICS
.IP \f3PM_ERR_MODE\f1
The
.I mode
parameter is invalid
.SH "SEE ALSO"
.BR pmcd (1),
.BR PMAPI (3),
.BR pmFetch (3),
.BR pmFetchArchive (3),
.BR pmGetArchiveEnd (3),
.BR pmGetArchiveLabel (3),
.BR pmGetInDom (3),
.BR pmLookupDesc (3),
.BR pmLookupInDom (3),
.BR pmLookupLabels (3),
.BR pmNameInDom (3)
and
.BR pmNewContext (3).

.\" control lines for scripts/man-spell
.\" +ok+ PM_TIME_XXXX
.\" +ok+ myarchive pmidlist mydelta mytime npmid mtime abs {from example C code}
.\" +ok+ pdf {from URL}


CHANGELOG
pcp-7.0.0 (1 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaopentelemetry: new exporter for OpenTelemetry metrics
	pmdarocestat: new RDMA over Converged Ethernet (RoCE) metrics
	pmdahdb: new agent exporting SAP HANA metrics
	pmdalinux: additional TCP networking metrics
	pmdalinux: additional kernel memory and hugetlb metrics
	pmdalinux: support Amazon Linux within kernel.uname.distro
	pmdaopenmetrics: improvements to dynamic metric name handling
	pmdaopenmetrics: allow configuration files with epoch times
	pmdapodman: fix container name/command handling in new podman
	pmdahacluster: additional pacemaker node metrics
	pmdasmart: add smart.wwid.* metrics with drive UUID for indom
	pmdammv: support empty instance domains and dynamic map resets
	pmdammv: export instance domain help text when requested
    - Server-side utilities and log management scripts:
	pmlogger, pmproxy: support for a remote archive push webhook;
	  providing central logging without server-side configuration
	pmproxy_daily: new daily script for pmlogger-pushed archives
	pmproxy: support both OpenMetrics and OpenTelemetry /metrics
	pmproxy: configurable HTTP headers (e.g. X-Grafana-Device-ID)
	pmlogger_daily: default compression is a dynamic zst/xz choice
	rc_pcp: legacy init.d pmcd+pmlogger wrapper script removed
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp.so.4: default PMAPI supports nanosecond precision times
	libpcp.so.3: provides backwards compatible libpcp interfaces
	pmapi.h: backwards compatible APIs via -DPMAPI_VERSION=2|4
	libpcp: flag for pmproxy to open archives at the last volume
	libpcp_import: protect against a second call to pmiEnd
	python api: fix (de)reference of a tuple in refresh callback
	python api: removed python version 2 support, 3.6 is minimum
    - Client tools and utilities:
	pcp-rocestat: client to report NVIDIA RoCE metrics
	pcp-dstat: fix the --time-adv option with archives
	pcp-htop: sync-up with latest htop.dev code fixes
	pcp-htop: improve pmcd disconnect/restart handling
	pmrep: fix iostat merge percentage formulas
	pmrep: allow for not preserving discrete metric values
	pmlogpush: new tool for pushing existing archives to pmproxy
	pcp2opentelemetry: new tool exporting to OpenTelemetry format
	pmseries_import: new tool for archive analysis container
	archive analysis container: removed rebuild of PCP sources
	archive analysis container: refactored to use real packages
	archive analysis container: set Grafana time window from archive
    - Build, infrastructure and packaging updates:
	build: removed Qt4 support from code base, 5.6 is minimum
	build: various improvements for reproducible builds
	build: archive-analysis container reworked to use packages
	build: fixes to better support parallel building
	build: fix multilib installs for pcp-libs-devel packages
	build: provide libvirt PMDA on any el7 if py3 modules exist
	build: clean up debian missing Build-Depends: packages
	selinux: update policy for AMD and NVIDIA gpu pmdas
	selinux: additional policy needed for pcp_pmie_t using ps
	selinux: add permissions allowing proc_psi_t access
	tmpfiles.d: additions to support better ostree integration
	pcp-zeroconf: use systemd presets to drive service start
    - Documentation:
	man pages: improved function prototypes in SYNOPSIS section
	man pages: updates to PCPCompat(1) describing major changes

pcp-6.3.8 (11 April 2025)
    - libpcp_web: fix selection and testing of exact match filtering
    - pmproxy: add After:valkey to the pmproxy systemd unit file
    - build: only regenerate debian/rules for Makepkgs builds
    - build: use the conditional build-me makefile target for pmdabpf
    - docs: update output format in pcp2elasticsearch man page
    - docs: fix webapi man page links, no longer hosted on pcp.io

pcp-6.3.7 (27 March 2025)
    - pmdadenki: RAPL MSR support
    - pmdabind2: updates for more recent bind versions
    - pmdalinux: additrional mptcp networking metrics
    - pcp2elasticsearch: support dynamic mapping
    - pmstat: avoid a corner-case infinite loop
    - libpcp: additional locking in __pmLogChangeVol
    - libpcp, pmproxy: PM_CTXFLAG_LAST_VOLUME pmNewContext flag
    - libpcp_web, pmseries: dynamic --load request rate adjusting
    - libpcp_web, pmseries: fix small memleak on archive load
    - pmproxy: REST API metric filtering for scrape requests
    - build: configure provides a manual pmdamysql override
    - build: configure always packages desktop icons in pcp-gui
    - build: debian packaging fixes and lintian improvements
    - build: fix unreliable configure.ac tests on scandir/dirent
    - build: use the vendor-supplied libinih package if available
    - selinux: several policy updates
    - docs: fix dup section in pmdaredis(1) man page

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time

qa/src/pmsetmode.c
/*
 * Copyright (c) 1995-2001 Silicon Graphics, Inc.  All Rights Reserved.
 * Copyright (c) 2024 Ken McDonell.  All Rights Reserved.
 *
 * Examples from pmSetMode(3) man page.
 *
 * NOTE: quick and dirty, no error checking
 */

#include <pcp/pmapi.h>
#include "libpcp.h"
#include <pcp/archive.h>

static pmLongOptions longopts[] = {
    PMOPT_ARCHIVE,	/* -a */
    PMOPT_DEBUG,	/* -D */
    PMOPT_INTERVAL,	/* -t */
    PMOPT_TIMEZONE,	/* -Z */
    PMOPT_HOSTZONE,	/* -z */
    PMAPI_OPTIONS_END
};

static pmOptions opts = {
    .short_options = "a:D:t:Z:z?",
    .long_options = longopts,
    .short_usage = "[options] ...",
};

int
main(int argc, char **argv)
{
    int		c;
    char	**namelist = NULL;
    pmID	*pmidlist = NULL;
    int		numpmid = 0;
    int		sts;

    struct timespec	mytime;
    struct timespec	mydelta = { 10, 0 };
    pmLogLabel		label;
    pmResult	*result;

    pmSetProgname(argv[0]);

    while ((c = pmGetOptions(argc, argv, &opts)) != EOF) {
	;
    }

    if (opts.flags & PM_OPTFLAG_EXIT) {
	pmflush();
	pmUsageMessage(&opts);
	exit(0);
    }

    if (opts.narchives != 1) {
	fprintf(stderr, "%s: need one -a option\n", pmGetProgname());
	pmUsageMessage(&opts);
	exit(EXIT_FAILURE);
    }

    if (opts.optind == argc) {
	fprintf(stderr, "%s: need at least one metric\n", pmGetProgname());
	pmUsageMessage(&opts);
	exit(EXIT_FAILURE);
    }

    if (opts.errors) {
	pmUsageMessage(&opts);
	exit(EXIT_FAILURE);
    }

    if (opts.interval.tv_sec != 0 || opts.interval.tv_nsec != 0)
	mydelta = opts.interval;

    /* metrics are argv[opts.optind] ... argv[argc-1] */
    while (opts.optind < argc) {
	namelist = (char **)realloc(namelist, (numpmid+1)*sizeof(namelist[0]));
	namelist[numpmid++] = argv[opts.optind];
	opts.optind++;
    }
    pmidlist = (pmID *)malloc(numpmid*sizeof(pmidlist[0]));

    if ((sts = pmNewContext(PM_CONTEXT_ARCHIVE, opts.archives[0])) < 0) {
	fprintf(stderr, "pmNewContext: %s: %s\n", opts.archives[0], pmErrStr(sts));
	exit(1);
    }
    if ((sts = pmLookupName(numpmid, (const char **)namelist, pmidlist)) < 0) {
	fprintf(stderr, "pmLookupName: %s ...: %s\n", namelist[0], pmErrStr(sts));
	exit(1);
    }
    free(namelist);

    /* EXAMPLE #1 */
    printf("Forwards ...\n");
    if ((sts = pmGetArchiveLabel(&label)) < 0) {
	fprintf(stderr, "pmGetArchiveLabel: %s\n", pmErrStr(sts));
	exit(1);
    }
    mytime = label.start;
    if ((pmSetMode(PM_MODE_INTERP, &mytime, &mydelta)) < 0) {
	fprintf(stderr, "pmSetMode: start: %s\n", pmErrStr(sts));
	exit(1);
    }
    while ((sts = pmFetch(numpmid, pmidlist, &result)) >= 0) {
	__pmDumpResult(stdout, result);
	pmFreeResult(result);
    }
    if (sts != PM_ERR_EOL)
	fprintf(stderr, "pmFetch: forward: %s\n", pmErrStr(sts));

    /* EXAMPLE #2 */
    printf("\nBackwards ...\n");
    if ((pmGetArchiveEnd(&mytime)) < 0) {
	fprintf(stderr, "pmGetArchiveEnd: %s\n", pmErrStr(sts));
	exit(1);
    }
    if ((pmSetMode(PM_MODE_BACK, &mytime, NULL)) < 0) {
	fprintf(stderr, "pmSetMode: end: %s\n", pmErrStr(sts));
	exit(1);
    }
    while ((sts = pmFetch(numpmid, pmidlist, &result)) >= 0) {
	__pmDumpResult(stdout, result);
	pmFreeResult(result);
    }
    if (sts != PM_ERR_EOL)
	fprintf(stderr, "pmFetch: backward: %s\n", pmErrStr(sts));

    free(pmidlist);

    return 0;
}

